src/LSST/control/ControlMain.cpp

```diff
@@ -57,16 +66,161 @@ ControlMain::Ptr ControlMain::getPtr() {
     return _thisPtr;
 }
 
-ControlMain& ControlMain::get() {
-    if (_thisPtr == nullptr) {
-        throw system::ConfigException(ERR_LOC, "ControlMain has not been setup.");
+ControlMain::ControlMain() {}
+
+ControlMain::~ControlMain() {}
+
+void ControlMain::run(int argc, const char* argv[]) {
+        thread t(&ControlMain::_cMain, this, argc, argv);
+        _mainThrd = move(t);
+}
+
+
+void signalHandler(int sig) {
+    if (sig == SIGPIPE) {
+        LWARN("Ignoring SIGPIPE sig=", sig);
+        return;
     }
-    return *_thisPtr;
+    LCRITICAL("Unhandled signal caught sig=", sig);
+    exit(sig);
 }
 
-ControlMain::ControlMain() {}
+void ControlMain::_cMain(int argc, const char* argv[]) {
+    LINFO("starting main");
+    util::Log& log = util::Log::getLog();
 
-ControlMain::~ControlMain() {}
+    // Get the configuration
+    system::Config& sysCfg = system::Config::get();
+
+    // Setup global items.
+    system::Globals::setup(sysCfg);
+
+    // Setup a simple signal handler to handle when clients closing connection results in SIGPIPE.
+    signal(SIGPIPE, signalHandler);
+
+    // Create the control system
+    _simCore.reset(new LSST::m2cellcpp::simulator::SimCore());
+    _simCore->start();
+    faultmgr::FaultMgr::setup();
+    control::FpgaIo::setup(_simCore);
+    control::MotionEngine::setup();
+    control::Context::setup();
+
+    // Register the PowerSystem with FpgaIo so that it gets updates.
+    auto powerSys = control::Context::get()->model.getPowerSystem();
+    control::FpgaIo::get().registerPowerSys(powerSys);
+
+    // Setup the control system
+    auto context = control::Context::get();
+    context->model.ctrlSetup();
+
+    // At this point, MotionCtrl and FpgaCtrl should be configured. Start the control loops
+    context->model.ctrlStart();
+
+    // At this point, the LabView code seems to want to put the system into
+    // ReadyIdle. We're NOT doing that. The system is going into StandbyState
+    // until it gets an explicit command to do something else.
+
+    // Start the telemetry server
+    // FUTURE: Telemetry for the GUI requires the ComControlServer->ComConnection::welcomeMsg to
+    //         work properly. Those elements should be in the telemetry so there's no need for a
+    //         ComControlServer connection for Telemetry. This has to wait until the existing
+    //         controller is no longer used.
+    LINFO("Starting Telemetry Server");
+    // FUTURE: get the correct entries into `system::Config`.
+    int const telemPort = 50001;
+    auto servTelemetryMap = system::TelemetryMap::Ptr(new system::TelemetryMap());
+    auto telemetryServ = system::TelemetryCom::create(servTelemetryMap, telemPort);
+
+    telemetryServ->startServer();
+    if (!telemetryServ->waitForServerRunning(5)) {
```

Comment:
What's so special about `5`? Also, what is the unit of that number? Is this the number of seconds, milliseconds, or microseconds? Consider using `chrono`.
The value may be also configured to avoid recompiling/rebuilding/redeploying the application each time the number needs to be adjusted.

Reply:
And, by the way, why have the separate "wait" method instead of passing the timeout to `startServer(5)` instead?

Reply:
There's nothing special about 5, other than even if the system is really busy, the connection should be ready in less than 5 seconds. Changed to
```
    int const secondsToWait = 5;
    if (!telemetryServ->waitForServerRunning(secondsToWait)) {
        // If it wasn't ready in 5 seconds, something is very wrong.
```

 The advantage of having it separate is that it can used in testing and called by other threads (most likely clients in tests) to make sure it is up before trying to connect.